### PR TITLE
[HGI-175] Register Segment as a destination for RETL

### PR DESCRIPTION
**Pull request by sayan-das-in**
This PR is to register Segment as a destination for RETL use-case.
This PR contains all Segment Specs except `alias`.

PRD: https://docs.google.com/document/d/1Iy-vqVxRQJXaRKgut2DZK10DwOmuq_Dm9z_a8aX1B7M/edit#heading=h.9aph5jk23l8w
JIRA Ticket: https://segment.atlassian.net/browse/HGI-175

## Testing
Tested using event tester in local environment.

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [ ] [Segmenters] Tested in the staging environment

### CODE REVIEW

File: packages/destination-actions/src/destinations/segment/index.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import type { DestinationDefinition } from '@segment/actions-core'
+import type { Settings } from './generated-types'
+import sendScreen from './sendScreen'
+import sendIdentify from './sendIdentify'
+import { SEGMENT_ENDPOINTS, DEFAULT_SEGMENT_ENDPOINT } from './properties'
+
+const destination: DestinationDefinition<Settings> = {
+  //Needs to be updated when name & slug are finalized
+  name: 'Segment',
```
Comment (brennan) [inline packages/destination-actions/src/destinations/segment/index.ts:9]: These look good.

File: packages/destination-actions/src/destinations/segment/index.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  extendRequest({ settings }) {
+    return {
+      headers: {
+        authorization: `Basic ${Buffer.from(settings.source_write_key + ':').toString('base64')}`
+      }
+    }
+  },
+  actions: {
+    sendScreen,
+    sendIdentify
```
Comment (prayansh) [inline packages/destination-actions/src/destinations/segment/index.ts:44]: any reason why we only support screen and identify?

Comment (prayansh) [reply packages/destination-actions/src/destinations/segment/index.ts:44]: if this is intended, then lets update the PR description to add that info

Comment (brennan) [reply packages/destination-actions/src/destinations/segment/index.ts:44]: Yep, we can update the PR description. This PR is just the first of many. The remaining will add support for additional Segment specced events - we'll support all but `alias`.

File: packages/destination-actions/src/destinations/segment/sendIdentify/index.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  },
+  perform: (request, { payload, settings }) => {
+    if (!payload.anonymous_id && !payload.user_id) {
+      throw new IntegrationError('Either Anonymous ID or User ID must be defined.', 'Misconfigured required field', 400)
+    }
+
+    const identifyPayload = {
+      userId: payload?.user_id,
+      annymousId: payload?.anonymous_id,
+      timestampe: payload?.timestamp,
```
Comment (NolanChan) [inline packages/destination-actions/src/destinations/segment/sendIdentify/index.ts:57]: typo i think! `timestamp`

File: packages/destination-actions/src/destinations/segment/sendIdentify/index.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    traits
+  },
+  perform: (request, { payload, settings }) => {
+    if (!payload.anonymous_id && !payload.user_id) {
+      throw new IntegrationError('Either Anonymous ID or User ID must be defined.', 'Misconfigured required field', 400)
+    }
+
+    const identifyPayload = {
+      userId: payload?.user_id,
+      annymousId: payload?.anonymous_id,
```
Comment (NolanChan) [inline packages/destination-actions/src/destinations/segment/sendIdentify/index.ts:56]: also typo here!

**Review by NolanChan — COMMENTED** [review]
Overall, logic for segment api is straightforward so that's nice :) But I found a few typo's on the field attributes, would it be possible to have some unit tests here?

File: packages/destination-actions/src/destinations/segment/sendIdentify/index.ts (hunk 3 above)
Comment (NolanChan) [reply packages/destination-actions/src/destinations/segment/sendIdentify/index.ts:57]: And i think an identical typo inside `screen`

File: packages/destination-actions/src/destinations/segment/index.ts (hunk 2 above)
Comment (sayan-das-in) [reply packages/destination-actions/src/destinations/segment/index.ts:51]: Updated the PR Description

File: packages/destination-actions/src/destinations/segment/sendIdentify/index.ts (hunk 3 above)
Comment (sayan-das-in) [reply packages/destination-actions/src/destinations/segment/sendIdentify/index.ts]: Fixed typos

File: packages/destination-actions/src/destinations/segment/sendIdentify/index.ts (hunk 4 above)
Comment (sayan-das-in) [reply packages/destination-actions/src/destinations/segment/sendIdentify/index.ts]: Fixed typos

**Comment by sayan-das-in** [comment]
> Overall, logic for segment api is straightforward so that's nice :) But I found a few typo's on the field attributes, would it be possible to have some unit tests here?

Added unit test cases and snapshot tests

File: packages/destination-actions/src/destinations/segment/sendGroup/__tests__/index.test.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    const responses = await testDestination.testAction('sendGroup', {
+      event,
+      mapping: defaultGroupMapping,
+      settings: {
+        source_write_key: 'test-source-write-key',
+        endpoint: DEFAULT_SEGMENT_ENDPOINT
+      }
+    })
+
+    expect(responses[0].status).toEqual(200)
```
Comment (NolanChan) [inline packages/destination-actions/src/destinations/segment/sendGroup/__tests__/index.test.ts:96]: In this test and similar tests, can we double check that we're receiving the attributes correctly? Thanks!

Comment (NolanChan) [reply packages/destination-actions/src/destinations/segment/sendGroup/__tests__/index.test.ts:96]: Examples of tests checking the `response.options.body` or `responses[0].options.body` can be found in fullstory, ga4, amplitude :)

Comment (sayan-das-in) [reply packages/destination-actions/src/destinations/segment/sendGroup/__tests__/index.test.ts:97]: Thanks for pointing out. I agree it's a good idea to add checks on the request body. Made changes to include additional checks.

**Review by NolanChan — APPROVED** [review]

**Pull request merged**